docs/user-guide/expressions/user-defined-functions.md

```diff
@@ -50,112 +20,114 @@ Use cases for `map_batches` in the `group_by` context are slim. They are only us
-
-```c
-"a" -> [10, 7]
-"b" -> [1]
-```
+We'll call `math.log()` on each individual value:
 
-If we would then apply a `shift` operation to the right, we'd expect:
+{{code_block('user-guide/expressions/user-defined-functions','individual_log',[])}}
 
-```c
-"a" -> [null, 10]
-"b" -> [null]
+```python exec="on" result="text" session="user-guide/udf"
+--8<-- "python/user-guide/expressions/user-defined-functions.py:individual_log"
 ```
 
-Let's try that out and see what we get:
+While this works, `map_elements()` has two problems:
 
-{{code_block('user-guide/expressions/user-defined-functions','shift_map_batches',[])}}
+1. **Limited to individual items:** Often you'll want to have a calculation that needs to operate on the whole `Series`, rather than individual items one by one.
+2. **Performance overhead:** Even if you do want to process each item individually, calling a function for each individual item is slow; all those extra function calls add a lot of overhead.
 
-```python exec="on" result="text" session="user-guide/udf"
---8<-- "python/user-guide/expressions/user-defined-functions.py:shift_map_batches"
-```
+Let's start by solving the first problem, and then we'll see how to solve the second problem.
 
-Ouch.. we clearly get the wrong results here. Group `"b"` even got a value from group `"a"` 😵.
+## Processing a whole `Series` with `map_batches()`
 
-This went horribly wrong because `map_batches` applied the function before aggregation, due to the `is_elementwise=True` parameter being provided. So that means the whole column `[10, 7, 1]` got shifted to `[null, 10, 7]` and was then aggregated.
+We want to run a custom on the contents of a whole `Series`.
+For demonstration purposes, let's say we want to calculate the difference between the mean of a `Series` and each value.
 
-So my advice is to never use `map_batches` in the `group_by` context unless you know you need it and know what you are doing.
+We can use the `map_batches()` API to run this function on either the full `Series` or individual groups in a `group_by()`:
 
-## To `map_elements`
+{{code_block('user-guide/expressions/user-defined-functions','diff_from_mean',[])}}
 
-Luckily we can fix previous example with `map_elements`. `map_elements` works on the smallest logical elements for that operation.
+```python exec="on" result="text" session="user-guide/udf"
+--8<-- "python/user-guide/expressions/user-defined-functions.py:diff_from_mean"
+```
 
-That is:
+## Fast operations with user-defined functions
 
-- `select context` -> single elements
-- `group by context` -> single groups
+The problem with a pure-Python implementation is that it's slow.
+In general, you want to minimize how much Python code you call if you want fast results.
+Calling a Python function for every `Series` isn't usually a problem, unless the `group_by()` produces a very large number of groups.
+However, running the `for` loop in Python, and then summing the values in Python, will be very slow.
```

Comment:
not sure what you mean by "every `Series`"

but I think these last two sentences can be removed, the first two ("The problem with a pure-Python ... In general ...") should be enough

Reply:
Was talking about the `Series` created by grouping, but yeah it's probably too much detail.